content/multi-level-enums.md

```diff
@@ -0,0 +1,213 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: implementation
+status: waiting-for-implementation
+presip-thread: https://contributors.scala-lang.org/t/pre-sip-foo-bar/9999
+title: SIP-NN - Multi-Level Enums
+---
+
+# SIP: Multi-Level Enums
+
+**By: Raphael Bosshard**
+
+## History
+
+| Date       | Version            |
+| ---------- | ------------------ |
+| 2025-09-15 | Initial Draft      |
+
+
+
+## Summary
+
+This proposal adds minimal syntax to allow enums with nested enumerations, maintaining full exhaustivity checking while keeping syntax clean and intuitive.
+
+## Motivation
+
+Scala 3 introduced `enum` as a concise and type-safe way to define algebraic data types. However, it currently supports only flat enums. Many real-world use cases, such as domain modeling or UI state machines, naturally require **hierarchical enums**, where cases are grouped into logical families or categories.
+
+Consider modeling animals:
+
+```scala
+enum Animal:
+  case Dog, Cat, Sparrow, Penguin
+```
+
+This flat structure works, but grouping `Dog` and `Cat` under `Mammal`, and `Sparrow` and `Penguin` under `Bird` is more expressive and enables cleaner abstraction.
+
+
+## Proposed solution
+
+### High-level overview
+
+Allow `enum` cases to contain **nested `enum` definitions**, using a consistent indentation-based syntax.
+
+```scala
+enum Animal:
+  case enum Mammal:
+    case Dog, Cat
+  case enum Bird:
+    case Sparrow, Pinguin
+```
+
+Each nested `enum` case defines a group of related subcases. The **nested enum is itself a valid case** of the parent enum, and its members are **also valid cases**, allowing full exhaustivity and pattern matching.
+
+### Specification
+
+#### Enum Definition:
+
+```scala
+enum Animal:
+  case enum Mammal:
+    case Dog, Cat
+  case enum Bird:
+    case Sparrow, Pinguin
+  case Fish
+```
+
+- `case enum Mammal:` introduces a **sub-enum case**.
+- Nested cases (`Dog`, `Cat`) are **automatically part of the parent enum** (`Animal`), as well as part of the sub-enum (`Mammal`).
+
+#### Desugaring / Type Relationships
+
+The above syntax desugars to an enum tree with subtype relationships:
+
+```scala
+sealed trait Animal
+
+object Animal:
+  sealed trait Mammal extends Animal
+  object Mammal:
+    case object Dog extends Mammal 
```

Comment:
to avoid getting into the weeds, because this is the wrong desugaring, its probably better to explain it in a set of recursive steps, where each `case enum` becomes an `enum` in the companion that extends the parent enum class (because an enum is a class not a sealed trait)
see the spec here: https://scala-lang.org/files/archive/spec/3.4/05-classes-and-objects.html#lowering-of-enum-definitions

Reply:
Thanks for the pointer!

I'll look into that and update the spec.